src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializerHelpers.cs

```diff
@@ -0,0 +1,364 @@
+                ODataPath odataPath = null;
+
+                if (internalRequest != null && urlHelper != null)
+                {
+                    string serviceRoot = urlHelper.CreateODataLink(
+                        internalRequest.Context.RouteName,
+                        internalRequest.PathHandler,
+                        new List<ODataPathSegment>());
+
+                    try
+                    {
+                        ODataUriParser parser = new ODataUriParser(readContext.Model, new Uri(serviceRoot), resourceWrapper.ResourceBase.Id);
+                        odataPath = parser.ParsePath();
+                    }
+                    catch (ODataException)
+                    {
+                        return null;
+                    }
+                    catch (UriFormatException)
+                    {
+                        return null;
+                    }
+                }
+
+                if (odataPath != null)
+                {
+                    return new Routing.ODataPath(odataPath);
+                }
+            }
+
+            Routing.ODataPath path = readContext.Path;
+
+            if (path == null)
+            {
+                return null;
+            }
+
+            IEdmEntityType entityType = path.EdmType.AsElementType() as IEdmEntityType;
+
+            if (entityType != null && path.EdmType.TypeKind == EdmTypeKind.Collection)
+            {
+                // create the uri for the current object, using path and key values
+                List<KeyValuePair<string, object>> keys = new List<KeyValuePair<string, object>>();
+                foreach (IEdmStructuralProperty keyProperty in entityType.Key())
+                {
+                    string keyName = keyProperty.Name;
+                    ODataProperty property = resourceWrapper.ResourceBase.Properties.Where(p => p.Name == keyName).FirstOrDefault();
+                    if (property == null && !readContext.DisableCaseInsensitiveRequestPropertyBinding)
+                    {
+                        //try case insensitive
+                        List<ODataProperty> candidates = resourceWrapper.ResourceBase.Properties.Where(p => String.Equals(p.Name, keyName, StringComparison.OrdinalIgnoreCase)).ToList();
+                        property = candidates.Count == 1 ? candidates.First() : null;
+                    }
+
+                    object keyValue = property?.Value;
+                    if (keyValue == null)
+                    {
+                        // Note: may be null if the payload did not include key values,
+                        // but still need to add the key so the path is semantically correct.
+                        // Key value type is not validated, so just use string.
+                        // Consider adding tests to ODL to ensure we don't validate key property type in future.
+                        keyValue = "Null";
+                    }
```

Comment:
There's a problem with this mitigation because an assumption is being made that the particular key property is of type string => "Null".

Reply:
Maybe we should give up trying to create the Uri when we encounter such a scenario or if was possible (and I doubt if it is) set `keyValue` to `default(Type?`)`

Reply:
the `keyValue` here is the keyValue object -- like new KeyValuePair<string, object>(keyName, keyValue). I don't think we should stop creating the `ODtaPath` if the keyValues were not passed in the payload.

Reply:
@ElizabethOkerio How do you settle of "Null" as a good fallback? Why not 0 for example? I'm not saying 0 is correct, I'm just wondering what makes "Null" the right token to come to the rescue...

Reply:
Can you add a test to verify that nothing breaks when we fallback on "Null"?